### ISSUE TITLE
TS Diagram should extend isopycnal lines to higher salinity

### ISSUE THREAD

**Issue by j-harbin** [dankelley/oce#1716]
As shown in the figure below, it would be useful to have isopycnal lines to higher salinity

![Screen Shot 2020-07-30 at 10 45 45 AM](https://user-images.githubusercontent.com/58750538/88930619-17ff9b00-d252-11ea-8caf-1ca6dab1a25c.png)

**Comment by dankelley** [comment]
Thanks. I'll have a look in the code to see where I'm setting the limits on what is drawn.  Possibly, though -- as we are discussing in a virtual meeting as I type this -- it's actually a *good* thing that we do not draw isopycnals into an unphysical  realm of salinity.

I'm not too sure what the physical realm is.  I'll try to see if I can find online PDFs of the original UNESCO docs.

**Comment by dankelley** [comment]
FYI, this works for the 'unesco' EOS, since
```R
library(oce)
plotTS(as.ctd(c(20,80),c(2,80),c(0,0)),eos="unesco")
```
yields
![image](https://user-images.githubusercontent.com/99469/88960809-73de1a00-d27a-11ea-8e2c-a063e5906785.png)

However, it fails with 'gsw' EOS, since
```R
 plotTS(as.ctd(c(20,80),c(2,80),c(0,0),lat=40,lon=-60),eos="gsw")
```
yields
![image](https://user-images.githubusercontent.com/99469/88960911-98d28d00-d27a-11ea-97b2-d59ac19c084a.png)

I have not looked too deeply in the code, but I suspect that one of the `gsw::gsw_` functions is giving `NA` values when things are out-of-range.  My next step is to look into the `gsw` code.  Then I'll document the limitation.

And I do think it might make sense to indicate the "valid" range for the "unesco" type, so if I can find the limitations in a UNESCO doc, I'll considering adding some pink lines or something.  (An underlay might be slightly tricky because I think we can draw TS diagrams on top of other things, etc.)

**Comment by dankelley** [comment]
The UNESCO doc
<img width="979" alt="Screen Shot 2020-07-30 at 4 44 18 PM" src="https://user-images.githubusercontent.com/99469/88967341-0b943600-d284-11ea-965f-76628d9ba37d.png">

has the following on page 23

<img width="766" alt="Screen Shot 2020-07-30 at 4 43 53 PM" src="https://user-images.githubusercontent.com/99469/88967312-fc14ed00-d283-11ea-8c09-514483d4d1a3.png">

**Comment by dankelley** [comment]
As suggested by @richardsc, I am inclined to make a new arg for `plotTS()` called `showValidRange` to default to TRUE, and that this would mean to draw the valid range in a red box. 

Overlays get confusing, because we draw the isopycnals after the data -- a subtle point that nevertheless is a thing I think we don't want to change, lest user's code produces results that they are not expecting.  And so an overlay would either have to be hatched (running the risk of hiding plot symbols) or using colour with transparency (which won't work on all plot devices).  I *could* draw the whole plot range at the very start (before drawing anything else) as a light gray or something, and then draw on top of that with `col=par("bg")` to set up the plot.  This might be tricky with the `inset` arg, though ... all of this makes me prefer to simply draw a red box.

I'll think about the aesthetics.

And I'll look up the gsw limits, too.

Likely I'll do this tomorrow or Saturday.

**Comment by dankelley** [comment]
The foundational document for gsw computation of density is McDougall et al. (2003), and section 3 of that paper outlines what the authors call the "funnel" of validity (which is more complex than simple ranges of S and T, or SA and CT).  There, I read that there is a cutoff of SA=40.  This value does not agree with the cutoff of (by eye) 50 g/kg shown in the TS diagram above.  So, I looked in the matlab code and found the definition of the funnel at https://github.com/TEOS-10/GSW-Matlab/blob/5723c4c54af8e8e25aa06d5ef8527843eb7cebbd/Toolbox/library/gsw_infunnel.m#L79, which reads as follows.  So that's a limit of 42 g/kg.

Thus, we seem to have three different upper limits for SA.

I'll continue to look into this, but gsw certainly appears to be more complicated than unesco.

```
in_funnel = ones(size(SA));

in_funnel(p > 8000 |...
    SA < 0 |...
    SA > 42 |...
    (p < 500 & CT < gsw_CT_freezing(SA,p)) |...
    (p > 500 & p < 6500 & SA < p*5e-3 - 2.5) |...
    (p > 500 & p < 6500 & CT > (31.66666666666667 - p*3.333333333333334e-3)) | ...
    (p > 6500 & SA < 30) |...
    (p > 6500 & CT > 10.0)) = 0;

in_funnel(isnan(SA) | isnan(CT) | isnan(p)) = NaN;
```

**References**

1. McDougall, Trevor J., David R. Jackett, Daniel G. Wright, and Rainer Feistel.
“Accurate and Computationally Efficient Algorithms for Potential Temperature
and Density of Seawater.” Journal of Atmospheric and Oceanic Technology 20, no.
5 (May 1, 2003): 730–41.
https://doi.org/10.1175/1520-0426(2003)20<730:AACEAF>2.0.CO;2.

**Comment by dankelley** [comment]
I am going to take some detailed notes here, as I look through the code.  Note that I will edit this comment in-place instead of adding new comments, so this will become an explanatory essay if I figure this out.  The links are permalinks to the code (in "develop" branch) at the moment of writing this.

In [`plotTS()`](https://github.com/dankelley/oce/blob/497862efb4b9e3c19d133956a00d95ebbaea364c/R/ctd.R#L4581)  within R/ctd.R, we trace the isopycnal curves based on temperature, density and pressure (set to 0) by using [`swSTrho()`](https://github.com/dankelley/oce/blob/497862efb4b9e3c19d133956a00d95ebbaea364c/R/sw.R#L673) within R/sw.R.  And, within [`swSTrho()`](https://github.com/dankelley/oce/blob/497862efb4b9e3c19d133956a00d95ebbaea364c/R/sw.R#L698) we compute that salinity for the gsw case using [`gsw::gsw_SA_from_rho`](https://github.com/TEOS-10/GSW-R/blob/29c1f62fc7d780f25d1e2071d627c492f90a01e4/R/gsw.R#L4255).

**Comment by dankelley** [comment]
From the previous two comments, I now know why `plotTS()` cuts off at SA=50, but I am not especially keen on assigning any real meaning to this, other than that this is a limit imposed by the underlying GSW code.  I will document this cutoff, and likely I will also mention that GSW papers, code, and documents suggest 3 different limits (that I know of so far).  I'll also document the limits for `eos="unesco"`.

**Comment by dankelley** [comment]
**VOTING**

A remaining question is what to do about indicating the range.  I propose as below, and am asking co-developers to vote yes or know with thumbs up or down (and to supply additional comments, if they want) prior to 1200h noon next Tuesday.  (I want to clean this up so @j-harbin and I can make a video on Thursday showing a TS diagram for argo data, before and after using QC to remove junk values.)

*PROPOSAL* `plotTS()` should limit the S and T range over which isopycnal curves are drawn, for the `eos="unesco"` case.  It would show these curves only for -2<T< 40C and 0<S<40.  A new argument should be added to `plotTS()` to get the present output, which involves (aphysical) extrapolation of the UNESCO formulae even to physically unrealistic values (e.g. the formula is happy to give densities at 300C).  That argument could be called `extrapolateIsopycnals`, or something like that, with default value `FALSE`.  Of course, the docs would explain all of this, and would recapitulate some of what I've said in previous comments about what at first glance seems to be contradictions in the gsw case (which I assume result from refinements of the code/formulae over the years).

**Comment by j-harbin** [comment]
I give a +1 to this idea. I like the idea of allowing the user to see that the data is unlikely (by not showing the isopycnal), while still allowing them to have the full isopycnal lines if desired.

**Comment by dankelley** [comment]
Given an explicit +1 vote by @j-harbin (and her help during a virtual meeting earlier today), given my own +1 vote, and given that @richardsc suggested the idea, I decided not to wait until next week and have instead  gone ahead and implemented a solution in "develop" commit 299763fe211c473ddf1c31fa1bc93c2c53f7f19b, for which the following code yields as shown in the snapshot.

It was a bit tricky deciding on defaults for the `eos="gsw"` case, and so I made what I thought was a reasonable choice, and discussed others in the docs for `?plotTS`.

With this done, I will now close this issue, with thanks to the other co-developers for their input.


```R
library(oce)
if (!interactive()) png("issue1716.png")
par(mfrow=c(2, 2))
ctd <- as.ctd(c(20,55),c(2,50),c(0,0),longitude=-63,latitude=43)
cex <- 1
plotTS(ctd, eos="unesco")
mtext("eos=\"unesco\", trimIsopycnals=TRUE", cex=cex, col=2, line=1, family="Times")
plotTS(ctd, eos="gsw")
mtext("eos=\"gsw\", trimIsopcynals=TRUE", cex=cex, col=2, line=1, family="Times")

plotTS(ctd, eos="unesco", trimIsopycnals=FALSE)
mtext("eos=\"unesco\", trimIsopycnals=FALSE", cex=cex, col=2, line=1, family="Times")
plotTS(ctd, eos="gsw", trimIsopycnals=FALSE)
mtext("eos=\"gsw\", trimIsopycnals=FALSE", cex=cex, col=2, line=1, family="Times")
if (!interactive()) dev.off()
```
yields as follows.  

![issue1716](https://user-images.githubusercontent.com/99469/89061974-08f51780-d33c-11ea-8518-5a682c8fed15.png)
